server_addon/applications/server/__init__.py

```diff
@@ -129,6 +148,85 @@ async def create_applications_attribute(self) -> bool:
             bool: 'True' if an attribute was created or updated.
         """
 
+        need_restart = await self.create_required_attributes()
+        await self.update_enums()
+        return need_restart
+
+    async def create_required_attributes(self) -> bool:
+        """Make sure there are required 'applications' and 'tools' attributes.
+        This only checks for the existence of the attributes, it does not populate
+        them with any data. When an attribute is added, server needs to be restarted,
+        while adding enum data to the attribute does not require a restart.
+        Returns:
+            bool: 'True' if an attribute was created or updated.
+        """
+
+        # keep track of the last attribute position (for adding new attributes)
+        apps_present = False
+        tools_present = False
+        apps_attribute_data = self._get_applications_def()
+        apps_attrib_name = apps_attribute_data["name"]
+
+        tools_attribute_data = self._get_tools_def()
+        tools_attrib_name = tools_attribute_data["name"]
+
+        last_index = -1
+        async for row in Postgres.iterate(
+            "SELECT name, position FROM attributes ORDER BY position"
+        ):
+            # check if the required attributes are present
+            # (in that case, we don't need to add them)
+            # also keep track of the last attribute position
+            if row["name"] == apps_attrib_name:
+                apps_present = True
+            elif row["name"] == tools_attrib_name:
+                tools_present = True
+            last_index = row["position"]
+
+        attributes_to_create = {}
+        if not apps_present:
+            attributes_to_create[apps_attrib_name] = {
+                "scope": apps_attribute_data["scope"],
+                "data": {
+                    "title": apps_attribute_data["title"],
+                    "type": apps_attribute_data["type"],
+                    "enum": [],
+                }
+            }
+
+        if not tools_present:
+            attributes_to_create[tools_attrib_name] = {
+                "scope": tools_attribute_data["scope"],
+                "data": {
+                    "title": tools_attribute_data["title"],
+                    "type": tools_attribute_data["type"],
+                    "enum": [],
+                },
+            }
+
+        # when any of the required attributes are not present, add them
+        # and return 'True' to indicate that server needs to be restarted
+        needs_restart = False
+        for name, payload in attributes_to_create.items():
+            await Postgres.execute(
+                "INSERT INTO attributes (name, position, scope, data) VALUES ($1, $2, $3, $4)",
+                name,
+                last_index + 1,
```

Comment:
Is there no better way in Postgres to ensure atomicity on an insert like this, ensuring the index is set after the current highest value on the actual insert itself? Or do we just not care because this 'runs on the server'?